taichi/backends/opengl/opengl_api.cpp

```diff
@@ -239,20 +239,24 @@ GLProgram *compile_glsl_program(std::string source) {
   return program;
 }
 
-GLSSBO *root_ssbo;
+namespace {
+  GLSSBO *root_ssbo;
```

Comment:
How about `std::unique_ptr<GLSSBO> root_ssbo`?

Reply:
+0xffffffff. I cannot stress enough how important it is to stay away from raw pointers *as much as possible* when that pointer actually owns the memory. Right now I can easily spot memory leaks, e.g. 

https://github.com/taichi-dev/taichi/blob/b865805452e542f61fa937933738a346f64d679e/taichi/codegen/codegen_opengl.cpp#L58

because `glsl` is never `delete`d. (This is also sad because the previous PRs were huge, and it is very hard to uncover all the problems in huge PRs)

Sticking with smart pointers can easily avoid such bugs. Without them, we are just waiting for bugs to shoot ourselves.

If you ever need to write `new` or `delete` on your own, think again. Most of the times they are just not the right design.